File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.java
class WordNode {
    Map<Character, WordNode> children = new HashMap();
    boolean isWord = false;
    public WordNode() {}
}

class WordDictionary {
    WordNode storage;

    public WordDictionary() {
        storage = new WordNode();
    }
    
    public void addWord(String word) {
        WordNode pointer = storage;
        
        for (char ch : word.toCharArray()) {
            if (!pointer.children.containsKey(ch)) {
                pointer.children.put(ch, new WordNode());
            }
            pointer = pointer.children.get(ch);
        }
        pointer.isWord = true;
    }
     
    
    public boolean search(String word) {
        return searchInNode(word, storage);
        
    }
    
    private boolean searchInNode(String word, WordNode node) {
        for (int i = 0; i < word.length(); ++i) {
            char ch = word.charAt(i);
            if (ch == '.') {
                for (char x : node.children.keySet()) {
                    WordNode child = node.children.get(x);
                    if (searchInNode(word.substring(i + 1), child)) {
                        return true;
                    }
                }
                return false;
            }
            if (!node.children.containsKey(ch)) {
                return false;
            }
            node = node.children.get(ch);
        }
        return node.isWord;
    }
}

File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.js
var WordDictionary = function() {
    this.Storage = {};
};

/** 
 * @param {string} word
 * @return {void}
 */
WordDictionary.prototype.addWord = function(word) {
    let pointer = this.Storage;
    for (let char of word) {
        if (!pointer[char]) { pointer[char] = {}}
        pointer = pointer[char];
    }

    pointer.end = true;
};

/** 
 * @param {string} word
 * @return {boolean}
 */
WordDictionary.prototype.search = function(word) {
    const recusionF = (string, storage) => {
        let pointer = storage;
        for (let i = 0; i < string.length; ++i) {
            const char = string[i];
            if (char === '.') {
                for (entry in pointer) {
                    if (recusionF(string.substring(i + 1), pointer[entry])) {
                        return true;
                    }
                }
                return false;
            }
            if (!pointer[char]) {
                return false;
            }
            pointer = pointer[char];
        }
        return !!pointer.end;
    }

    return recusionF(word, this.Storage)
};

/** 
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = new WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */